### DISCUSSION TITLE
Is it possible to ignore specific undefined names?

### DISCUSSION THREAD

**Discussion by zendynar** [astral-sh/ruff#7033] [Q&A] [answered]
First of all, thanks a lot for this great project, it helps me a lot in my work!

I use ruff as an extension for VSCode/VSCodium. I work with an ETL tool that support python scripting, so the code will run in that environment instead of my local machine. I typically copy the code locally to be checked with ruff and other tools.

That tool provides some modules that are somehow available in the environment, so I don't have to do `import`. But when I copy the code to check it, ruff provides a ton of errors about undefined name:

![image](https://github.com/astral-sh/ruff/assets/105803717/86f0d188-4879-4caf-a677-aaf614a02d7f)

I could disable the rule undefined-name (F821)[#](https://beta.ruff.rs/docs/rules/undefined-name/#undefined-name-f821) altogether but I still want to catch all other undefined name errors for all other names other than the specific one (in my case `api`).

My workaround for now is to manually insert `import api` into the file, but I have to delete it afterwards as otherwise my tool will throw an error. Is there a way to tell ruff to only ignore one name and still check the others. If I use `requests` I still want to be notified if I forgot to `import requests`.  I know I can use `noqa` but this is not sustainable as I use `api` too many times. I haven't found a way to only exclude one type of name. Is there a convenient solution I can use?

**Comment by RextConnor** [comment]
theres probably another solution. Visual Studio Code still would NEED to have the module downloaded in order for it to work. If that all is working you'd need to specify what the python requests is going to do. I'd start off with atleast doing something where it would return the result.

**Reply by zendynar** [reply]
@RextConnor, thanks for your message. I know the code won't work in VSCode without the module. In fact even if I could download the module the code will not run locally. I only use VSCode to use ruff and other tools for code formatting and validation and then send it over to work on the remote system

**Comment by MichaReiser** [comment, marked as answer]
You can use the [`builtins` settings](https://beta.ruff.rs/docs/settings/#builtins) to define additional "environment provided" builtins. See this [Playground](https://play.ruff.rs/e3f0961a-e308-419e-ae26-73579c7da9dd) for an example.
